cmd/proxy/actions/catalog.go

```diff
@@ -18,16 +19,15 @@ type catalogRes struct {
 	NextPageToken   string                `json:"next,omitempty"`
 }
 
-// RegisterHandlers is a convenience method that registers
-// all the storage routes.
-func RegisterHandlers(r *mux.Router, s Backend) {
-	r.Handle("/catalog", CatalogHandler(s)).Methods(http.MethodGet)
-}
-
-// CatalogHandler implements GET baseURL/catalog
-func CatalogHandler(s Backend) http.Handler {
+// catalogHandler implements GET baseURL/catalog
+func catalogHandler(s storage.Backend) http.HandlerFunc {
 	const op errors.Op = "download.CatalogHandler"
+	cs, isCataloger := s.(storage.Cataloger)
```

Comment:
I am not a huge fan of type assertion and finding out if a type has been implemented or not. When we had the conversation, I thought it was already implemented it this way.

Is this how go recommends doing it? If so, can you show me, I would like to know more.

Reply:
@manugupt1 interface-assertion is actually a good pattern if used correctly, as an example look at https://golang.org/pkg/net/http/#Flusher (in other words, the main http.ResponseWriter interface *does* have the Flush method, but only optionally, so you use interface-assertion to find out)

Reply:
yeah! the issue that I have right now is that I am not sure which way is better, both seem very equivalent to me. 

Also, if you think about it, we have already implemented these functions in the storage: 
https://github.com/fedepaol/athens/commit/0258e17d8905fa851c78f8e8ed2e73358135f816#diff-b646c51b29c85b8406f5792043c69de7R12

and isCataloger will never be false (with the code that we have written so far); which makes it functionally equivalent for the user but still uses the underlying interfaces.

Another thing, is error types, right now we have ServerError and NotImplemented. I think there are storages that won't be able to support catalog which should have a NotSupported error. See here: https://github.com/gomods/athens/issues/996

With this implementation, in my mind, we are increasing the probablity that a contributor is not able to identify the implementation is missing and misses opportunity to contribute to  our repo.

Reply:
@manugupt1 

>Also, if you think about it, we have already implemented these functions in the storage:
I have removed those from this PR, therefore they'll come back as false. That's the whole point, we don't want to provide non-implemented implementations :) 

>I think there are storages that won't be able to support catalog
is exactly the point of this PR. 

>With this implementation, in my mind, we are increasing the probability that a contributor is not able to identify the implementation is missing and misses opportunity to contribute to our repo.

We want people to have an easier time to contribute, so the smaller the interface the quicker they can PR something. The Catalog method is complex due to pagination and backend implementation, and so it should definitely be optional. 

Lastly `NotImplemented` and `NotSupported` are too similar, no need to be that verbose, we can revisit this later on.

Reply:
azureblob.Cataloger?

Reply:
I missed azureblob, LGTM